### PR TITLE
Python 3.10 update.

### PR DESCRIPTION
**Pull request by dumol**
Scope
-----

Update Python.

Changes
-------

Use latest Python 3.10. (Python 3.11 builds fail because of https://github.com/python/cpython/issues/101668.)

The `libedit` patch was removed, as it's not worth porting it to apply to 3.10 sources only for building on macOS…

**Drive-by changes**:
  * Removed 32bit Windows build.
  * Added an arm64 macOS build.
  * Build the Windows/macOS packages on whatever GitHub supports as `-latest` (currently Windows Server 2022 and macOS 12).
  * Build the generic Linux package on Amazon 2 with glibc 2.26.
  * Updated the other libs to build, where applicable: SQLite, libffi, xz.
  * Updated `cryptography` to newly-released version 40.0.1.
  * Build and check for PyNaCl.
  * Check for `cffi` too.
  * Collect and check for `MarkupSafe`, `charset-normalizer`, and `pywin32` (the latter only on Windows).

Testing
-------

Review changes and the automated test results.

Try the resulting test packages, e.g. https://bin.chevah.com:20443/testing/3.10.11.5cdf347/

### CODE REVIEW

**Comment by dumol** [comment]
Anything blocking this?

**Comment by adiroiban** [comment]
I am commenting here:

Here are the missing extra binary wheels.

For py3.8 on Linux, I have updated some pre-compiled versions , but we need them for Alpine.

We might just remove support for alpine for the initial py3, and later we can look into adding these packages

Saved ./bcrypt-4.0.1-cp36-abi3-manylinux_2_28_x86_64.whl
Saved ./cryptography-40.0.1-cp36-abi3-manylinux_2_28_x86_64.whl
Saved ./cffi-1.15.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
Saved ./PyNaCl-1.5.0-cp36-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl

**Comment by dumol** [comment]
[…]

> Saved ./bcrypt-4.0.1-cp36-abi3-manylinux_2_28_x86_64.whl

`bcrypt` 4.0.1 is already present on both Linux builds.


> Saved ./cryptography-40.0.1-cp36-abi3-manylinux_2_28_x86_64.whl 

Updated `cryptography` to newly-released version 40.0.1 through https://github.com/chevah/pythia/pull/39/commits/9eb343cb2afb961a24d6646f4c4e14423b826815.


> Saved ./cffi-1.15.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl 

`cffi` 1.15.1 already present on all platforms, including Alpine. Now also checked through https://github.com/chevah/pythia/pull/39/commits/ef91078c5e286cca3af23fe7c8b986f49b1deb1a.


> Saved ./PyNaCl-1.5.0-cp36-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl

PyNaCl is now built and checked for through https://github.com/chevah/pythia/pull/39/commits/7bc9f35f2204dd9c94561561fedf613f7cb974e0.

**Comment by dumol** [comment]
If you want a Python 3.8.x release with everything but the Python 3.10 update, let me know, I can do such a release independently of this update to Python 3.10.

**Comment by adiroiban** [comment]
I think that python 3.8 is ok.

But is important to make sure that for 3.10 we have the binary packages.

Another binary package is https://pypi.org/project/MarkupSafe/1.1.1/#files

I will continue to add the packages here

**Comment by adiroiban** [comment]
Another binary package is https://pypi.org/project/charset-normalizer/2.0.11/

Note that for now we use 2.0.11 since it works on both py2 and py3.

**Comment by adiroiban** [comment]
Basically, for py3 on both linux and Windows, we should look into our existing pypi.chevah.com files and review any file that ends with:

* win_amd64.whl  - windows binary
* manylinux1_x86_64.whl (or similar) - linux binary 
* macosx_10_6_intel.whl (or similar)

Now, our pypi.chevah.com already has some binary package, since I used this as a quick fix,
but for the long term, these packages should be provided by pythia.

**Comment by adiroiban** [comment]
Another thing.
For now we need `--use-deprecated=legacy-resolver` from pip, so this most probably means that we can only use pip 21.0

**Comment by adiroiban** [comment]
on py3.8 there is another issue on Windows.

Somehow, the site packates are not autoamtically enabled.

I can see the `lib/python38._pth` which has the `site.main()` disabled.

I can patch as part of our deps system, but it should be enabled in pythia by defautl.

```
$ cat build-py3/lib/python38._pth
python38.zip
.

# Uncomment to run site.main() automatically
#import site
lib\site-packages
```

**Comment by adiroiban** [comment]
Another thing that should be done for py3.10


we no longer need the `-3` check

```diff
diff --git a/pythia.sh b/pythia.sh
index baa9dd52e..a3ad6a61a 100755
--- a/pythia.sh
+++ b/pythia.sh
@@ -106,7 +106,7 @@ check_source_folder() {
 # Called to trigger the entry point in the virtual environment.
 # Can be overwritten in pythia.conf
 execute_venv() {
-    ${PYTHON_BIN} $PYTHON3_CHECK -c 'from paver.tasks import main; main()' "$@"
+    ${PYTHON_BIN} -c 'from paver.tasks import main; main()' "$@"
 }
 
 
@@ -905,15 +905,6 @@ if [ "$COMMAND" == "deps" ] ; then
     install_base_deps
 fi
 
-case $COMMAND in
-    test_ci|test_py3)
-        PYTHON3_CHECK='-3'
-        ;;
-    *)
-        PYTHON3_CHECK=''
-        ;;
-esac
-
```

File: src/Python-Windows/chevahbs [hunk 1]
```diff
@@ -32,4 +32,4 @@ echo "## Extracting ZIP archive to ${INSTALL_DIR}... ##"
 execute unzip -q "$ARCHIVE" -d "${INSTALL_DIR}"
 
 echo "## Adding site-packages to python*._pth file... ##"
-echo "lib\\site-packages" >> ${INSTALL_DIR}/python38._pth
+echo "lib\\site-packages" >> ${INSTALL_DIR}/python310._pth
```
Comment (adiroiban) [inline src/Python-Windows/chevahbs:35]: I don't this this is needed.

The `lib/site-packages` should be added by default

**Comment by dumol** [comment]
> Another thing. For now we need `--use-deprecated=legacy-resolver` from pip, so this most probably means that we can only use pip 21.0

Am I missing something? Documentation for pip 23.0.1 still shows this parameter as present: https://pip.pypa.io/en/stable/development/release-process/#use-deprecated.

**Comment by adiroiban** [comment]
>  Am I missing something? Documentation for pip 23.0.1 still shows this parameter as present: https://pip.pypa.io/en/stable/development/release-process/#use-deprecated.

True. it look like this feature was not removed .. yet
It still works with 23.0.1
So all good for pip.

**Comment by adiroiban** [comment]
Regarding the python wheel binary packages.

It's ok to have them here.

The only hard requirement  is for musl/alpine.

For windows and generic linux, I can find them already pre-compiled on the official PyPI site.

File: src/chevah-python-test/test_python_binary_dist.py [hunk 2]
```diff
@@ -464,17 +496,42 @@ def main():
...
-            exit_code = 152
+            exit_code = 171
         else:
             print('ctypes %s' % (ctypes.__version__,))
 
+        try:
+            import win32service
+            win32service.EnumWindowStations()
+        except:
+            sys.stderr.write('"pywin32" missing or broken.\n')
```
Comment (adiroiban) [inline src/chevah-python-test/test_python_binary_dist.py:532]: We don't really use win32service in the code.

win32security security should be a better test, and we should show the error in the logs.

```suggestion
            import win32security
            win32security.win32api.GetCurrentThread()()
        except Exception as error:
            sys.stderr.write('"pywin32" missing or broken.\n {}'.format(error))
```

I will try this patch.

File: .github/workflows/bare.yaml [hunk 3]
```diff
@@ -23,7 +21,7 @@ concurrency:
 # Also increase timeout-minutes for the relevant OS when debugging remotely.
 # Version 3 of mxschmitt/action-tmate should also work on Windows.
 env:
-  TMATE_DEBUG: 'no'
+  TMATE_DEBUG: 'yes'
```
Comment (adiroiban) [inline .github/workflows/bare.yaml:24]:
```suggestion
  TMATE_DEBUG: 'no'
```

File: .github/workflows/bare.yaml [hunk 4]
```diff
@@ -37,7 +35,7 @@ jobs:
       fail-fast: false
       matrix:
         runs-on: [ windows-2022, windows-2019 ]
-    timeout-minutes: 15
+    timeout-minutes: 60
```
Comment (adiroiban) [inline .github/workflows/bare.yaml:38]:
```suggestion
    timeout-minutes: 15
```

File: src/Python-Windows/chevahbs (hunk 1 above)
Comment (adiroiban) [inline src/Python-Windows/chevahbs:35]: enable site-packages side effects 

```suggestion
echo "import site; site.main()" >> ${INSTALL_DIR}/python310._pth
```

**Review by adiroiban — COMMENTED** [review]
try some changes

File: src/Python-Windows/chevahbs [hunk 5]
```diff
@@ -32,4 +32,4 @@ echo "## Extracting ZIP archive to ${INSTALL_DIR}... ##"
 execute unzip -q "$ARCHIVE" -d "${INSTALL_DIR}"
 
 echo "## Adding site-packages to python*._pth file... ##"
-echo "lib\\site-packages" >> ${INSTALL_DIR}/python38._pth
+echo "import site; site.main()" >> ${INSTALL_DIR}/python310._pth
```
Comment (adiroiban) [inline src/Python-Windows/chevahbs:35]:
```suggestion
echo "site.main()" >> ${INSTALL_DIR}/python310._pth
```

File: src/Python-Windows/chevahbs [hunk 6]
```diff
@@ -32,4 +32,4 @@ echo "## Extracting ZIP archive to ${INSTALL_DIR}... ##"
 execute unzip -q "$ARCHIVE" -d "${INSTALL_DIR}"
 
 echo "## Adding site-packages to python*._pth file... ##"
-echo "lib\\site-packages" >> ${INSTALL_DIR}/python38._pth
+echo "site.main()" >> ${INSTALL_DIR}/python310._pth
```
Comment (adiroiban) [inline src/Python-Windows/chevahbs:35]:
```suggestion
echo "site.main()" >> ${INSTALL_DIR}/python310._pth
echo "lib\\site-packages" >> ${INSTALL_DIR}/python38._pth
```

File: src/Python-Windows/chevahbs [hunk 7]
```diff
@@ -32,4 +32,5 @@ echo "## Extracting ZIP archive to ${INSTALL_DIR}... ##"
...
+echo "lib\\site-packages" >> ${INSTALL_DIR}/python10._pth
```
Comment (adiroiban) [inline src/Python-Windows/chevahbs:36]:
```suggestion
echo "lib\\site-packages" >> ${INSTALL_DIR}/python310._pth
```

File: src/Python-Windows/chevahbs [hunk 8]
```diff
@@ -32,4 +32,5 @@ echo "## Extracting ZIP archive to ${INSTALL_DIR}... ##"
 execute unzip -q "$ARCHIVE" -d "${INSTALL_DIR}"
 
 echo "## Adding site-packages to python*._pth file... ##"
-echo "lib\\site-packages" >> ${INSTALL_DIR}/python38._pth
+echo "site.main()" >> ${INSTALL_DIR}/python310._pth
```
Comment (adiroiban) [inline src/Python-Windows/chevahbs:35]:
```suggestion
echo "import site" >> ${INSTALL_DIR}/python310._pth
echo "site.main()" >> ${INSTALL_DIR}/python310._pth
```

File: src/chevah-python-test/test_python_binary_dist.py [hunk 9]
```diff
@@ -464,17 +501,42 @@ def main():
...
         except:
             sys.stderr.write('"ctypes - windll" is missing.\n')
-            exit_code = 152
+            exit_code = 171
         else:
             print('ctypes %s' % (ctypes.__version__,))
 
+        try:
+            import win32security
+            win32security.win32api.GetCurrentThread()
```
Comment (adiroiban) [inline src/chevah-python-test/test_python_binary_dist.py:535]:
```suggestion
            win32security.GetCurrentThread()
```

File: src/chevah-python-test/test_python_binary_dist.py [hunk 10]
```diff
@@ -4,6 +4,11 @@
 import sys
 import platform
 import subprocess
+import site
+
+# FIXME
+# I have no idea why site.main is not enabled at this point.
+site.main()
```
Comment (adiroiban) [inline src/chevah-python-test/test_python_binary_dist.py:11]:
```suggestion
```

File: .github/workflows/bare.yaml (hunk 3 above)
Comment (adiroiban) [inline .github/workflows/bare.yaml:24]:
```suggestion
  TMATE_DEBUG: 'no'
```

File: .github/workflows/bare.yaml [hunk 11]
```diff
@@ -36,8 +34,8 @@ jobs:
       # Workflow won't be cancelled at the first failed job.
       fail-fast: false
       matrix:
-        runs-on: [ windows-2022, windows-2019 ]
-    timeout-minutes: 15
+        runs-on: [ 'windows-2022' ]
+    timeout-minutes: 60
```
Comment (adiroiban) [inline .github/workflows/bare.yaml:38]:
```suggestion
        runs-on: [ 'windows-2019' ]
    timeout-minutes: 15
```

File: src/chevah-python-test/test_python_binary_dist.py [hunk 12]
```diff
@@ -464,17 +496,42 @@ def main():
...
         except:
             sys.stderr.write('"ctypes - windll" is missing.\n')
-            exit_code = 152
+            exit_code = 171
         else:
             print('ctypes %s' % (ctypes.__version__,))
 
+        try:
+            import win32security
+            win32security.GetCurrentThread()
```
Comment (adiroiban) [inline src/chevah-python-test/test_python_binary_dist.py:530]:
```suggestion
            import win32api
            win32api.GetCurrentThread()
```

**Comment by adiroiban** [comment]
Also, it would help a lot to just run 64bit on Windows, since then we will no longer have to execute in pythia.sh the `systeminfo.exe` helper... it just slows down the whole thing

File: .github/workflows/bare.yaml [hunk 13]
```diff
@@ -36,7 +34,7 @@ jobs:
       # Workflow won't be cancelled at the first failed job.
       fail-fast: false
       matrix:
-        runs-on: [ windows-2022, windows-2019 ]
+        runs-on: [ 'windows-2019' ]
```
Comment (adiroiban) [inline .github/workflows/bare.yaml:37]: Try 64bit 
```suggestion
        runs-on: [ 'windows-2022' ]
```

File: src/Python-Windows/chevahbs (hunk 1 above)
Comment (dumol) [reply src/Python-Windows/chevahbs:37]: Looking at those backslashes, maybe it's only needed on Windows?

**Comment by dumol** [comment]
> on py3.8 there is another issue on Windows.
> 
> Somehow, the site packates are not autoamtically enabled.

Is this solved now with your latest changes?

**Comment by dumol** [comment]
Thanks for all the fixes! Is there anything left to improve? I'm a bit dizzy after checking all the commits done through suggestions above… :-]

Also, as discussed on Friday on #chevah, what Python modules should I remove from Pythia? I see `charset_normalizer` is only available as a binary wheel. Other stuff? 

You mentioned `setproctitle` and `MarkupSafe`, but the former is built from sources on platforms other than Windows, while the latter is build on Alpine.

There is also `cryptography` which is not built anywhere lately, but I guess you want this included, so I'm not sure where to draw the line…

File: src/Python-Windows/chevahbs [hunk 14]
```diff
@@ -32,4 +32,6 @@ echo "## Extracting ZIP archive to ${INSTALL_DIR}... ##"
 execute unzip -q "$ARCHIVE" -d "${INSTALL_DIR}"
 
 echo "## Adding site-packages to python*._pth file... ##"
-echo "lib\\site-packages" >> ${INSTALL_DIR}/python38._pth
+echo "import site" >> ${INSTALL_DIR}/python310._pth
+echo "site.main()" >> ${INSTALL_DIR}/python310._pth
+echo "lib\\site-packages" >> ${INSTALL_DIR}/python310._pth
```
Comment (adiroiban) [inline src/Python-Windows/chevahbs:37]: Let's try without it

```suggestion
```

**Comment by dumol** [comment]
On a second thought, if we don't include `cryptography`, then it can be installed and packaged with SFTPPlus as needed. Which would save us from re-releasing Pythia whenever there is a `cryptography` security update… Maybe useful after all?

File: .github/workflows/bare.yaml [hunk 15]
```diff
@@ -36,7 +34,7 @@ jobs:
       # Workflow won't be cancelled at the first failed job.
       fail-fast: false
       matrix:
-        runs-on: [ windows-2022, windows-2019 ]
+        runs-on: [ 'windows-2022' ]
```
Comment (adiroiban) [inline .github/workflows/bare.yaml:37]: maybe remove the matrix from here

**Comment by dumol** [comment]
> Another thing that should be done for py3.10
> 
> we no longer need the `-3` check

Done through https://github.com/chevah/pythia/pull/39/commits/f2a18c5e583a25d41a75d167707bfa27bfd6bb01.

File: .github/workflows/bare.yaml (hunk 15 above)
Comment (dumol) [reply .github/workflows/bare.yaml]: Done through https://github.com/chevah/pythia/pull/39/commits/7d5e2275b13f021120c2e4a4e5687b0805707f16.

**Comment by dumol** [comment]
Went through all the comments once more, let me know if I missed something before approving this. Thanks!

**Comment by dumol** [comment]
For the record, I've tried building the generic Linux package on Amazon Linux 1 (aka 2018.03), which has the same glibc version as RHEL 7. And which has gcc 7.2 available as an extra package.

The package builds, but it has the same issues as on CentOS 7, it links against `/lib64/libfreebl3.so`, which belongs to an irremovable system package.

**Comment by adiroiban** [comment]
I am leaving a comment here, but most probably needs to be done in a separate ticket.

For the Windows version, besides the lib/python310.zip we also need to distribute the source code.

Without the source code, the python works, but debugging is hard.

So for Windows, besides the current embeddable version, we need to find a way to also get the source code avaiable inside the build folder

**Comment by dumol** [comment]
> So for Windows, besides the current embeddable version, we need to find a way to also get the source code avaiable inside the build folder

Sure, no problem, I'll fill in a new ticket… But what sources you want included? The win package of python-package only has some include files in `/python2.7-win-x64/lib/include/`… You want more? All the source code of Python? Only some parts of it?

What kind of debugging you have in mind? Debugging Python binaries?

**Comment by dumol** [comment]
For the record, merged into master as discussed on #chevah last week on Friday.

**Pull request merged**